File: app/javascript/controllers/admin/lti_databases_controller.js
import * as Rbase from "@app_root/app/javascript/rbase_common.js"
import { RbaseController } from "@app_root/app/javascript/rbase_stimulus.js"
import { get, post, put, patch, destroy } from '@rails/request.js'

export default class extends RbaseController {
  index() {
    super.index();
    if (Rbase.getParams('clear') == 'true') {
      Rbase.clearWebStorageFormValue('lti_database');
    }
  }
  
  new() {
    super.new();
  }
  
  edit() {
    super.edit();
  }
  show() {
    super.show();
    Rbase.showFormDisbaled();
  }
  
  async button_update_pem(event) {
    console.log("lti_databases_controller.js->button_update_pem()");
    event.preventDefault();
    event.stopPropagation();

    var self = this;
    if (window.confirm(self.paramsValue.build_confirm_message)) {
      const response = await get(self.paramsValue.url1, {
        contentType: "application/json"
      });
      
      if (response.ok) {
        console.log("success");
        const data = await response.json;
        $("#lti_database_private_key_file").html(data.pem);
        $("#lti_database_public_key").html(data.pub_pem);
        return false;
      }
    }
    return false;
  }
  
  async button_update_kid(event) {
    console.log("lti_databases_controller.js->button_update_kid()");
    event.preventDefault();
    event.stopPropagation();

    var self = this;
    if (window.confirm(self.paramsValue.build_confirm_message)) {
      const response = await get(self.paramsValue.url2, {
        contentType: "application/json"
      });
      
      if (response.ok) {
        const data = await response.json;
        console.log("success:"+data.kid);
        $("#lti_database_kid").val(data.kid);
        return false;
      }
    }
    return false;
  }
}

File: app/controllers/admin/lti_databases_controller.rb
module Admin
  class LtiDatabasesController < CustomAdminApplicationController
    before_action :set_lti_database, only: [:show, :edit, :update, :destroy]
    before_action :setup_values, only: [:index, :new, :create, :edit, :update, :show]

    respond_to :html
    def index
      if params[:lti_databases_search_conditions]
        @condition = LTIDatabases::SearchConditions.new(search_condition_params)
        @condition.sites = Array(@condition.sites).map(&:to_s)
        return render 'index' unless @condition.valid?
        @condition.current_admin_user = current_admin_user
        session[:lti_databases_search_conditions] = @condition
        
        @lti_databases = @condition.search.page(params[:page])
      else
        if session[:lti_databases_search_conditions]
          @condition = session[:lti_databases_search_conditions]
          @condition.current_admin_user = current_admin_user
          @lti_databases = @condition.search.page(params[:page])
        else
          @condition = LTIDatabases::SearchConditions.new
          @condition.current_admin_user = current_admin_user
          @lti_databases = @condition.search.page(params[:page])
        end
      end
    end

    def show
      respond_with(@lti_database)
    end

    def new
      @lti_database = LTIDatabase.new
      @stimulus_params = {
        url1: button_update_pem_admin_lti_databases_path,
        url2: button_update_kid_admin_lti_databases_path,
        update_confirm_message: I18n.t(:"views.common.update_confirm_message"),
        build_confirm_message: I18n.t(:"views.common.build_confirm_message")
      }.to_json
      respond_with(@lti_database)
    end

    def edit
      @stimulus_params = {
        url1: button_update_pem_admin_lti_databases_path,
        url2: button_update_kid_admin_lti_databases_path,
        update_confirm_message: I18n.t(:"views.common.update_confirm_message"),
        build_confirm_message: I18n.t(:"views.common.build_confirm_message")
      }.to_json
    end

    def create
      @lti_database = LTIDatabase.new(lti_database_params)

      if @lti_database.save
        flash[:notice] = t("views.common.create_complete_message")
        respond_with(@lti_database, location: admin_lti_databases_url)
      else
        @stimulus_params = {
          url1: button_update_pem_admin_lti_databases_path,
          url2: button_update_kid_admin_lti_databases_path,
          update_confirm_message: I18n.t(:"views.common.update_confirm_message"),
          build_confirm_message: I18n.t(:"views.common.build_confirm_message")
        }.to_json
        render :new, status: :unprocessable_entity
      end
    end
    def update
      ActiveRecord::Base.transaction do
        @lti_database.assign_attributes(lti_database_params)
        if @lti_database.save
          flash[:notice] = t("views.common.update_complete_message")
          respond_with(@lti_database, location: admin_lti_databases_url)
        else
          @stimulus_params = {
            url1: button_update_pem_admin_lti_databases_path,
            url2: button_update_kid_admin_lti_databases_path,
            update_confirm_message: I18n.t(:"views.common.update_confirm_message"),
            build_confirm_message: I18n.t(:"views.common.build_confirm_message")
          }.to_json
          render :edit, status: :unprocessable_entity
        end
      end
    end

    def destroy
      flash[:notice] = t("views.common.destroy_complete_message") if @lti_database.destroy
      respond_with(@lti_database, location: admin_lti_databases_url)
    end

    desc :auth_as => :index, :display_name => 'admin/lti_databases/button_update_pem'
    def button_update_pem
      pem, pub_pem = ::LTIDatabase.create_pem
      result = {
        pem: pem,
        pub_pem: pub_pem
      }
      render json: result
    end

    desc :auth_as => :index, :display_name => 'admin/lti_databases/button_update_pem'
    def button_update_kid
      kid = ::LTIDatabase.create_kid
      result = {
        kid: kid
      }
      render json: result
    end

    private
    def setup_values
      @sites = current_admin_user.sites.all
    end
    
    def set_lti_database
      @lti_database = LTIDatabase.find(params[:id])
    end

    def search_condition_params
      params[:lti_databases_search_conditions][:sites].to_a.reject!{|x|x.blank?}
      params.require(:lti_databases_search_conditions).permit!
    end
    
    def lti_database_params
      params.require(:lti_database).permit!
    end
    
  end
end
